### DISCUSSION TITLE
Custom User Action Tracking and Analysis

### DISCUSSION THREAD

**Discussion by ash-peng** [thoughtspot/cs_tools#155] [Q&A] [answered]
Hello CS Tools Devs,

We have a feature request that we don't know if it is feasible / within CS Tools capacities.

**Objective:**
Extract ThoughtSpot’s metadata to capture user actions and adoption metrics via CS Tools, which was recommended to us. Reconfigure this data to meet specific business needs (more details below) and push the transformed metadata back into ThoughtSpot to create tailored liveboards for various data owners.

**Challenge:**
ThoughtSpot captures comprehensive user activity data, but it is currently aggregated across all user bases. We need a method to isolate and separate these actions by specific data tables/worksheets. Future user groups will expand, and users may interact with multiple data sets, making it infeasible to isolate data by user IDs alone.

**Requirements:**
Split Data by User Groups: Separate user actions by data tables/worksheets to distinguish between different user bases (note that some user work on multiple categories so they belong to multiple user bases).
Create Monitoring Dashboards: Develop specific liveboards that visualize user metrics for each user group, ensuring stakeholders can monitor relevant data without overlap from other groups.

**Questions:**
Based on the above requirements,
Are there existing features or APIs in ThoughtSpot to facilitate detailed extraction and categorization of user actions by specific data objects?
What best practices or tools are recommended for achieving this level of detailed user action tracking?

Thank you.

**Comment by boonhapus** [comment, marked as answer]
Hey @ash-peng , we have a tool called **[Searchable](https://thoughtspot.github.io/cs_tools/tools/searchable/)** which can capture a lot of similar data that you're looking for. It comes with a few Liveboards to help guide usage of the underlying database schema.

The only visuals we have of this right now are **[in the 1.5.0 release notes](https://thoughtspot.github.io/cs_tools/changelog/1.5.0/#__tabbed_1_3)**, please take a look at the Liveboard and let me know if this is in line with what you're thinking.

This tool pulls from a collection of APIs, but primary the metadata/search and user security endpoints. We then reshape the data and build a transactional style data model from it for use in ThoughtSpot.

--

ThoughtSpot's BI Server worksheet doesn't have the lineage to be able to tie ad-hoc Searches to a worksheet, but the rest of the viz content (Answer, Liveboard) can be walked back in the dependency chain to see the source Worksheet.

The other interesting challenge you'll find is that ThoughtSpot has no concept of "primary" group. So if I user is a member of multiple groups, there's no reliable way to attribute their search activity to any particular group.

**Reply by ash-peng** [reply]
Thank you for your reply! The visuals look good and seems to be what we would be looking for - Do we just run the command to install the tool and we will be able to use it to build liveboards in ThoughtSpot? Or will the liveboard be built in there? Is there any way to test it beforehand, as we only have one production environment. Thanks again

**Reply by boonhapus** [reply]
Hey @ash-peng , good questions. I don't have documentation describing the full process yet, but here's the answer to your Qs..

---

CS Tools is a wrapper around the ThoughtSpot APIs. There's a command called `metadata` where we authenticate to ThoughtSpot and fetch data, transform it into tabular format, and then load it to any of the supported data stores on **[the Syncers page](https://thoughtspot.github.io/cs_tools/syncer/what-is/)**. There's also a command called `deploy` which will customize TML for the assets to-be-created in your ThoughtSpot, and then use the TML API to load it for you as well. I would recommend creating a separate connection for this data though, so it's easier to identify in your cluster.

As part of `deploy`, we don't share the ThoughtSpot content to anyone, so only your Admins should be able to see it by default. It's also not a "heavy" activity for the server, so there's no cause to worry about downtime or sluggishness of your ThoughtSpot cluster.

> Another question: Does CS Tools need to be constantly maintained?

This data fetching `metadata` activity should happen regularly, but the TML upload `deploy` only needs to be run once. We also do release updates every few weeks, as **[there's a suite of tools offered](https://thoughtspot.github.io/cs_tools/tools/)** and sometimes those need bugfixes or new features.

**Reply by ash-peng** [reply]
@boonhapus Thank you for this clear and specific response. Just to confirm, `metadata` fetches data from ThoughtSpot, then `deploy` deploys the metadata fetched, is this correct? 
    - More specifically, for our use case, we would use `metadata` first, then `deploy` to see the fetched metadata loaded into admin's ThoughtSpot, then run `metadata` regularly for CS Tools to fetch new data but not `deploy`, unless there is a release update.

I also noticed the `bi-server` command. User actions seem to be included inside `bi-server` according to the documentation? Or does `metadata` include all that's in `bi-server`? Could you point me towards more information for the three commands?

Thanks again for your answer.

**Comment by ash-peng** [comment]
Another question: Does CS Tools need to be constantly maintained? 

Thank you!

**Reply by boonhapus** [reply]
Answer this one **[along with your other Qs](https://github.com/thoughtspot/cs_tools/discussions/155#discussioncomment-9988159)** 👍
